The download client should be able to determine if a placeholder should be used

pvollan opened the pull request:
#### e3fcc5543b97e63cdd7dab7c6accbed7cbb6991d
<pre>
The download client should be able to determine if a placeholder should be used
<a href="https://bugs.webkit.org/show_bug.cgi?id=278737">https://bugs.webkit.org/show_bug.cgi?id=278737</a>
<a href="https://rdar.apple.com/134794671">rdar://134794671</a>

Reviewed by Alex Christensen.

The download client should be able to determine if a placeholder should be used for the final location of the
download. This patch prepares for this by passing the value from the UI process to the Networking process.

* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::publishProgress):
(WebKit::PendingDownload::didBecomeDownload):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKDownload.mm:
(-[_WKDownload publishProgressAtURL:]):
* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
(-[WKDownload setDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h: Added.
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::decideDestinationWithSuggestedFilename):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.h:
(WebKit::DownloadProxy::setUseDownloadPlaceholder):
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(TestWebKitAPI::DecidePlaceholderPolicy)):
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.mm:
(-[TestDownloadDelegate _download:decidePlaceholderPolicy:]):

Canonical link: <a href="https://commits.webkit.org/283043@main">https://commits.webkit.org/283043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fad6993143022b4d3cc126601f71ae95b0cbea5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17637 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69047 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15629 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67141 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15911 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52239 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/10794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68089 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32861 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/37711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/13649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14505 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/13987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70752 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8975 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13465 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59565 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9007 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/56338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59797 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7403 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9857 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/41023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegate.h:
```diff
@@ -95,6 +100,8 @@ WK_SWIFT_UI_ACTOR
  */
 - (void)download:(WKDownload *)download didFailWithError:(NSError *)error resumeData:(nullable NSData *)resumeData;
 
+- (void)download:(WKDownload *)download decidePlaceholderPolicy:(WK_SWIFT_UI_ACTOR void (^)(WKPlaceholderPolicy))completionHandler;
```
This needs WK_API_AVAILABLE

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKDownload.mm:
```diff
@@ -79,7 +79,11 @@ - (void)cancel
 - (void)publishProgressAtURL:(NSURL *)URL
 {
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    _download->_download->publishProgress(URL, false);
```
Could you give false a name?

achristensen07 commented on Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
```diff
@@ -164,9 +164,13 @@ void DownloadProxy::decideDestinationWithSuggestedFilename(const WebCore::Resour
+            m_client->decidePlaceholderPolicy(*this, [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), destination = WTFMove(destination)] (WebKit::UseDownloadPlaceholder usePlaceholder) mutable {
+                bool shouldEnableModernDownloadProgress = CFPreferencesGetAppBooleanValue(CFSTR("EnableModernDownloadProgress"), CFSTR("com.apple.WebKit"), nullptr);
+                if (shouldEnableModernDownloadProgress)
+                    publishProgress(URL::fileURLWithFileSystemPath(destination), usePlaceholder == WebKit::UseDownloadPlaceholder::Yes);
```
I think publishProgress needs to be part of the completion handler.  We shouldn't send the destination to the network process without the placeholder policy.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h:
```diff
@@ -0,0 +1,59 @@
+ @discussion The placeholder policy specifies whether a placeholder file should be created in
+ the Downloads directory when the download is in progress.
+ */
+- (void)download:(WKDownload *)download _decidePlaceholderPolicy:(WK_SWIFT_UI_ACTOR void (^)(_WKPlaceholderPolicy))completionHandler WK_API_AVAILABLE(WK_MAC_TBA, WK_IOS_TBA);
```
Underscore needs to be before the first part of the selector.  _download:decidePlaceholderPolicy:
SPI probably doesn't need WK_SWIFT_UI_ACTOR

pvollan replied:
Fixed in latest patch!

Thanks for reviewing!

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
```diff
@@ -97,6 +98,7 @@ class PendingDownload : public NetworkLoadClient, public IPC::MessageSender, pub
     URL m_progressURL;
 #if HAVE(MODERN_DOWNLOADPROGRESS)
     Vector<uint8_t> m_bookmarkData;
+    UseDownloadPlaceholder m_useDownloadPlaceholder { false };
```
UseDownloadPlaceholder::No

achristensen07 commented on Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
```diff
@@ -161,12 +161,16 @@ void DownloadProxy::decideDestinationWithSuggestedFilename(const WebCore::Resour
+        m_client->decidePlaceholderPolicy(*this, [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), destination = WTFMove(destination), sandboxExtensionHandle = WTFMove(sandboxExtensionHandle), allowOverwrite] (WebKit::UseDownloadPlaceholder usePlaceholder) mutable {
+            bool shouldEnableModernDownloadProgress = CFPreferencesGetAppBooleanValue(CFSTR("EnableModernDownloadProgress"), CFSTR("com.apple.WebKit"), nullptr);
+            if (!destination.isEmpty() && shouldEnableModernDownloadProgress)
+                publishProgress(URL::fileURLWithFileSystemPath(destination), usePlaceholder);
```
I think we should remove this and just add another parameter to the completionHandler instead.  That'll be a little tricky with _WKDownload still existing, but we'll soon remove _WKDownload's ability to publish progress.

achristensen07 reviewed (changes requested):
This needs a test that makes sure the delegate is actually called.

pvollan committed:
The download client should be able to determine if a placeholder should be used
https://bugs.webkit.org/show_bug.cgi?id=278737
rdar://134794671

Reviewed by Alex Christensen.

The download client should be able to determine if a placeholder should be used for the final location of the
download. This patch prepares for this by passing the value from the UI process to the Networking process.

* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.cpp:
(WebKit::PendingDownload::publishProgress):
(WebKit::PendingDownload::didBecomeDownload):
* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKDownload.mm:
(-[_WKDownload publishProgressAtURL:]):
* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
(-[WKDownload setDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h: Added.
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::decideDestinationWithSuggestedFilename):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.h:
(WebKit::DownloadProxy::setUseDownloadPlaceholder):
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(TestWebKitAPI::DecidePlaceholderPolicy)):
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.mm:
(-[TestDownloadDelegate _download:decidePlaceholderPolicy:]):

Canonical link: https://commits.webkit.org/283043@main
